internal/controller/dpuoperatorconfig_controller_test.go

```diff
@@ -0,0 +1,174 @@
+		return err
+	}
+
+	return nil
+}
+
+var _ = Describe("Main Controller", Ordered, func() {
+	var cancel context.CancelFunc
+	var ctx context.Context
+
+	BeforeAll(func() {
+		By("setting up env variables for tests")
+		err := os.Setenv("DPU_DAEMON_IMAGE", "mock-image")
+		Expect(err).NotTo(HaveOccurred())
+
+		By("create test DPU operator namespace")
+		namespace := &corev1.Namespace{
+			TypeMeta: metav1.TypeMeta{},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testNamespace,
+			},
+			Spec:   corev1.NamespaceSpec{},
+			Status: corev1.NamespaceStatus{},
+		}
+		Expect(k8sClient.Create(context.Background(), namespace)).Should(Succeed())
+
+		By("verify test DPU operator namespace is created")
+		retrieved_namespace := &corev1.Namespace{}
+		err = k8sClient.Get(context.Background(), client.ObjectKey{Name: testNamespace}, retrieved_namespace)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(retrieved_namespace.ObjectMeta.Name).To(Equal(testNamespace))
+
+		By("create DpuOperatorConfig controller k8s objs")
+		config := &configv1.DpuOperatorConfig{}
+		config.SetNamespace(testNamespace)
+		config.SetName(testDpuOperatorConfigName)
+		config.Spec = configv1.DpuOperatorConfigSpec{
+			Mode:     "host",
+			LogLevel: 2,
+		}
+		Expect(k8sClient.Create(context.Background(), config)).Should(Succeed())
+		DeferCleanup(func() {
+			err := k8sClient.Delete(context.Background(), config)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		By("verify DpuOperatorConfig controller k8s objs is created")
+		retrieved_config := &configv1.DpuOperatorConfig{}
+		err = k8sClient.Get(context.Background(), client.ObjectKey{Namespace: testNamespace, Name: testDpuOperatorConfigName}, retrieved_config)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(retrieved_config.ObjectMeta.Namespace).To(Equal(testNamespace))
+		Expect(retrieved_config.ObjectMeta.Name).To(Equal(testDpuOperatorConfigName))
+
+		By("setup controller manager")
+		k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+			Scheme: scheme.Scheme,
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		By("setup controller manager reconciler")
+		err = (&DpuOperatorConfigReconciler{
+			Client: k8sManager.GetClient(),
+			Scheme: k8sManager.GetScheme(),
+		}).SetupWithManager(k8sManager)
+		Expect(err).ToNot(HaveOccurred())
+
+		ctx, cancel = context.WithCancel(context.Background())
+
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer GinkgoRecover()
+			By("start controller manager")
+			err := k8sManager.Start(ctx)
+			Expect(err).ToNot(HaveOccurred())
+		}()
+
+		DeferCleanup(func() {
+			By("shut down controller manager")
+			cancel()
+			wg.Wait()
+		})
+
+		// HACK: Wait until controller manager boots up
+		// FIXME: Find a better way to detect that the controller manager is in a good state. Perhaps we will implement
+		// the "status" portion of the CR in the future.
+		time.Sleep(8 * time.Second)
+	})
+
```

Comment:
Could we restructure the tests in as follows:
- when the controller manager is up, creating a DpuOperatorConfig CR should cause a daemonset to be created
- When the controller manager is up, changing the DPUOperatorConfig CR should cause the daemonset to be udpated

I believe your next PR will add:
- When the controller manager is up and DpuOperatorConfig CR is created, the sriov DP is created too on the same nodes.